File: crc.c
/*Write a program to compute CRC code for the polynomials

sample
input=11100011
11001
11100011
output=Enter data to be transmitted:
Enter the Generating polynomial:
Data padded with n-1 zeros : 111000110000CRC or Check value is : 0000Final data
to be sent : 111000110000Enter the received data:
Data received: 11100011No error detected

*/
#include <stdio.h>
#include <string.h>

void xorOperation(char dividend[], char generator[], int n)
{

    for (int i = 0; i < n; i++)
    {

        if (dividend[i] == generator[i])
            dividend[i] = '0';
        else
            dividend[i] = '1';
    }
}

void crc(char data[], char generator[], char crcCode[])

{

    int dataLen = strlen(data);
    int genLen = strlen(generator);
    char temp[128];
    strcpy(temp, data);
    for (int i = dataLen; i < dataLen + genLen - 1; i++)
    {
        temp[i] = '0';
    }
    temp[dataLen + genLen - 1] = '\0';
    char dividend[128];
    strncpy(dividend, temp, genLen);

    for (int i = 0; i < dataLen; i++)
    {

        if (dividend[0] == '1')
        {
            xorOperation(dividend, generator, genLen);
        }
        for (int j = 0; j < genLen - 1; j++)
        {

            dividend[j] = dividend[j + 1];
        }
        dividend[genLen - 1] = temp[i + genLen];
    }

    strncpy(crcCode, dividend, genLen - 1);
    crcCode[genLen - 1] = '\0';
}

int main()
{
    char data[128], generator[128], crcCode[128], transmitted[128];

    printf("Enter data to be transmitted: ");
    scanf("%s", data);

    printf("\nEnter the Generating polynomial: ");
    scanf("%s", generator);

    printf("\nData padded with n-1 zeros : ");
    crc(data, generator, crcCode);
    printf("%s%s", data, crcCode);

    printf("\nCRC or Check value is : %s", crcCode);

    strcpy(transmitted, data);
    strcat(transmitted, crcCode);
    printf("\nFinal data to be sent : %s", transmitted);


    printf("\nEnter the received data: ");
    scanf("%s", transmitted);

    printf("\nData received: %s", transmitted);

    crc(transmitted, generator, crcCode);

    if (strcmp(crcCode, "0000") == 0)
        printf("\nNo error detected\n");
    else
        printf("\nError detected\n");

    return 0;
}

File: CRC.java
import java.util.Scanner;

public class CRC {

    static String data;
    static String generator;
    static String checkValue;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Input data and generator polynomial
        System.out.print("Enter data to be transmitted: ");
        data = scanner.nextLine();
        System.out.print("Enter the generator polynomial: ");
        generator = scanner.nextLine();

        // Append zeros to data based on generator length
        String paddedData = data + "0".repeat(generator.length() - 1);
        System.out.println("\nData padded with zeros: " + paddedData);

        // Calculate CRC
        checkValue = calculateCRC(paddedData);
        System.out.println("CRC value: " + checkValue);

        // Append CRC to original data
        String transmittedData = data + checkValue;
        System.out.println("Transmitted data: " + transmittedData);

        // Receiver checks for errors
        System.out.print("\nEnter received data: ");
        String receivedData = scanner.nextLine();
        String remainder = calculateCRC(receivedData);

        if (remainder.contains("1")) {
            System.out.println("Error detected in received data.");
        } else {
            System.out.println("No error detected in received data.");
        }

        scanner.close();
    }

    static String calculateCRC(String input) {
        StringBuilder result = new StringBuilder(input);

        for (int i = 0; i <= input.length() - generator.length(); i++) {
            if (result.charAt(i) == '1') {
                for (int j = 0; j < generator.length(); j++) {
                    result.setCharAt(i + j, result.charAt(i + j) == generator.charAt(j) ? '0' : '1');
                }
            }
        }

        return result.substring(input.length() - (generator.length() - 1));
    }
}